### PR TITLE
REV-022: create-listing-form

### PR DESCRIPTION
**Pull request by hosamAjlouni-op**
Implementing the (create listing form page)

### CODE REVIEW

**Review by mohdabed123 — APPROVED** [review]

**Comment by AhmadAlShwaiki** [comment]
Please merge the following PRs before this one.
1- https://github.com/abukhadraop-or/reverifi/pull/33
2-https://github.com/abukhadraop-or/reverifi/pull/32
3-https://github.com/abukhadraop-or/reverifi/pull/36

**Review by AhmadAlShwaiki — APPROVED** [review]

File: src/components/create-listing/FeatureSelection/index.jsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  Title,
+} from './feature-selection.styles';
+
+/**
+ * Feature Selection section of Create Listing page.
+ *
+ * @param {Set} featureIds Set of selected feature id's.
+ * @param {Function} handleFeatureClick feature click handler function.
+ * @return {JSX.Element}
+ */
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/FeatureSelection/index.jsx:19]:
```suggestion
/**
 * Feature Selection section of Create Listing page.
 *
 * @param {Set}      featureIds         Set of selected feature id's.
 * @param {Function} handleFeatureClick feature click handler function.
 *
 * @return {JSX.Element}
 */
```

File: src/components/create-listing/FeatureSelection/index.jsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  featureIds: PropTypes.objectOf(PropTypes.number),
+  /**
+   * feature click handler function.
+   */
+  handleFeatureClick: PropTypes.func,
+};
+
+FeatureSelection.defaultProps = {
+  featureIds: [],
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/FeatureSelection/index.jsx:52]: How `featureIds` have prop type `objectOf` and default props `Array`??

File: src/components/create-listing/FeatureSelection/index.jsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+FeatureSelection.propTypes = {
+  /**
+   * Set of selected feature id's.
+   */
+  featureIds: PropTypes.objectOf(PropTypes.number),
+  /**
+   * feature click handler function.
+   */
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/FeatureSelection/index.jsx:47]:
```suggestion

```

File: src/components/create-listing/FormInputFields/index.jsx [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+ * @param {Object} props                                 The component props.
+ * @param {string} [props.overviewText=null]             Overview input value.
+ * @param {Function} [props.register=null]               The react-hook-form register function.
+ * @param {Object} [props.errors=null]                   Error message/s.
+ * @param {Function} [props.handleIdentifierClick =null] The function to handle on click event for buttons.
+ * @param {Set} [props.selectedIdentifiers=null]         The currently selected buttons.
+ * @param {Function} [props.setOverviewText=null]        Update the value of the overviewText state.
+ *
+ * @return {JSX.Element}
+ */
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/FormInputFields/index.jsx:43]:
```suggestion
/**
 * Render the form input fields.
 *
 * @param {Object}   props                               The component props.
 * @param {string}   [props.overviewText=null]           Overview input value.
 * @param {Function} [props.register=null]               The react-hook-form register function.
 * @param {Object}   [props.errors=null]                 Error message/s.
 * @param {Function} [props.handleIdentifierClick =null] The function to handle on click event for buttons.
 * @param {Set}      [props.selectedIdentifiers=null]    The currently selected buttons.
 * @param {Function} [props.setOverviewText=null]        Update the value of the overviewText state.
 *
 * @return {JSX.Element}
 */
```

File: src/components/create-listing/FormInputFields/index.jsx [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+  errors,
+  handleIdentifierClick,
+  selectedIdentifiers,
+  setOverviewText,
+}) {
+  /**
+   *
+   * @type {number} Maximum characters allowed for overview input.
+   *
+   */
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/FormInputFields/index.jsx:56]:
```suggestion
  /**
   * @type {number} Maximum characters allowed for overview input.
   */
```

File: src/components/create-listing/FormInputFields/index.jsx [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+  const overviewCharsLimit = 140;
+
+  /**
+   * Updates overview input value.
+   *
+   * @param {Object} event On change event object.
+   *
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/FormInputFields/index.jsx:63]:
```suggestion
```

File: src/components/create-listing/FormInputFields/index.jsx [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+   */
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/FormInputFields/index.jsx:64]:
```suggestion
    /**
   * Updates overview input value.
   *
   * @param {React.ChangeEvent} event Object containing the new value for overview input.
   */```

File: src/components/create-listing/ListingFormMap/index.jsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import locationImg from 'assets/images/location.svg';
+import React from 'react';
+import { LocationContainer } from './listing-form-map.styles';
+/**
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/ListingFormMap/index.jsx:4]:
```suggestion

/**
```

File: src/components/create-listing/ListingImageInput/index.jsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+import FilesList from 'components/shared/FilesList';
+import FormFileInput from 'components/shared/FormFileInput';
+import {
+  Container,
+  ImageInputSection,
+  Wrapper,
+  InputInterface,
+  UploadIcon,
+} from './listingImageInput.styles';
+
```
Comment (Mahmoudzaid2994) [inline src/components/create-listing/ListingImageInput/index.jsx:12]: Add docs here please

File: src/components/shared/FilesList/index.jsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable react/no-array-index-key */
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FilesList/index.jsx:1]:
```suggestion

```

File: src/components/shared/FilesList/index.jsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  SelectedItemInnerSection,
+  SelectedItemsList,
+  ThumbnailWrapper,
+} from './filesList.styles';
+
+function FilesList({ files, onDelete }) {
+  return (
+    <SelectedItemsList>
+      <span>{files.length} Files Selected</span>
+      {files.map((file, i) => (
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FilesList/index.jsx:16]: you shouldn't use index as a key, try to pass props key and use it 'as example'

File: src/components/shared/FormCheckbox/index.jsx [hunk 12]
```diff
@@ -6,11 +6,11 @@ import { Container, Input, Label } from './form-checkbox.style';
...
- * @param {string} [props.id=null] Input ID.
- * @param {string} [props.label=null] Label text.
- * @param {string} props.name Input name.
- * @param {Function} [props.onChange] The on change event.
+ * @param {Object} props                   The component props.
+ * @param {string} [props.id=null]         Input ID.
+ * @param {string} [props.label=null]      Label text.
+ * @param {string} props.name              Input name.
+ * @param {Function} [props.onChange]      The on change event.
  * @param {Function} [props.register=null] The react-hook-form register function.
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FormCheckbox/index.jsx:14]:
```suggestion
 * @param {Object}   props                 The component props.
 * @param {string}   [props.id=null]       Input ID.
 * @param {string}   [props.label=null]    Label text.
 * @param {string}   props.name            Input name.
 * @param {Function} [props.onChange]      The on change event.
 * @param {Function} [props.register=null] The react-hook-form register function.
```

File: src/components/shared/FormFileInput/index.jsx [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Renders a form file input
+ *
+ * @param  {string}       acceptedType  Accepted files types.
+ * @param  {boolean}      multiple      Determines files quantity to single or many.
+ * @param  {Function}     setFiles      A function to set and update files.
+ * @param  {JSX.Element}  children      The child component as an interface for the file input.
+ *
+ * @return {JSX.Element}
+ */
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FormFileInput/index.jsx:14]:
```suggestion
/**
 * Renders a form file input
 *
 * @param {string}      acceptedType Accepted files types.
 * @param {boolean}     multiple     Determines files quantity to single or many.
 * @param {Function}    setFiles     A function to set and update files.
 * @param {JSX.Element} children     The child component as an interface for the file input.
 *
 * @return {JSX.Element}
 */
```

File: src/components/shared/FormFileInput/index.jsx [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+function FormFileInput({ acceptedTypes, multiple, onAddFiles, children }) {
+  const isFileTypeAccepted = (f) =>
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FormFileInput/index.jsx:16]:
```suggestion
  const isFileTypeAccepted = (file) =>
```

Also add docs for `file` props

File: src/components/shared/FormFileInput/index.jsx [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    !acceptedTypes || acceptedTypes.includes(f.type);
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FormFileInput/index.jsx:17]:
```suggestion
  const isFileTypeAccepted = (file) =>
    !acceptedTypes || acceptedTypes.includes(file.type);
```

Also add docs for `file` props

File: src/components/shared/FormFileInput/index.jsx [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    const acceptedFiles = [...e.target.files].filter((f) =>
+      isFileTypeAccepted(f)
+    );
+
+    if (!acceptedFiles.length) return;
+
+    onAddFiles(acceptedFiles);
+
+    e.target.value = null;
+  };
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FormFileInput/index.jsx:29]:
```suggestion
   const handleChange = (event) => {
    const acceptedFiles = [...event.target.files].filter((file) =>
      isFileTypeAccepted(file)
    );

    if (!acceptedFiles.length) return;

    onAddFiles(acceptedFiles);

    event.target.value = null;
  };
```

and also add docs

File: src/components/shared/FormInput/index.jsx [hunk 17]
```diff
@@ -5,16 +5,16 @@ import { Container, Error, Input, Label } from './form-input.styles';
...
- * @param {string} [props.id=null] Input ID.
- * @param {string} [props.label=null] Label text.
- * @param {string} props.name Input name.
- * @param {Function} [props.onChange] The on change event.
+ * @param {Object} props                    The component props.
+ * @param {string} [props.error=null]       Error message.
+ * @param {string} [props.id=null]          Input ID.
+ * @param {string} [props.label=null]       Label text.
+ * @param {string} props.name               Input name.
+ * @param {Function} [props.onChange]       The on change event.
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FormInput/index.jsx:13]:
```suggestion
 * @param {Object}   props                         The component props.
 * @param {string}   [props.error=null]            Error message.
 * @param {string}   [props.id=null]               Input ID.
 * @param {string}   [props.label=null]            Label text.
 * @param {node}     [props.labelIconElement=null] Label icon element.
 * @param {string}   props.name                    Input name.
```

File: src/components/shared/FormTextArea/index.jsx [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * @param {Object} props                    The component props.
+ * @param {string} [props.error=null]       Error message.
+ * @param {string} [props.id=null]          Input ID.
+ * @param {string} [props.label=null]       Label text.
+ * @param {string} props.name               Input name.
+ * @param {Function} [props.onChange]       The on change event.
+ * @param {string} [props.placeholder=null] Input placeholder.
+ * @param {Function} [props.register=null]  The react-hook-form register function.
+ * @param {boolean} [props.rounded=true]    Is input borderers rounded.
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FormTextArea/index.jsx:22]:
```suggestion
 * @param {Object}   props                         The component props.
 * @param {string}   [props.error=null]            Error message.
 * @param {string}   [props.id=null]               Input ID.
 * @param {string}   [props.label=null]            Label text.
 * @param {node}     [props.labelIconElement=null] Label icon element.
 * @param {string}   props.name                    Input name.
 * @param {Function} [props.onChange]              The on change event.
 * @param {string}   [props.placeholder=null]      Input placeholder.
 * @param {Function} [props.register=null]         The react-hook-form register function.
 * @param {boolean}  [props.rounded=true]          Is input borderers rounded.
```

File: src/components/shared/FormTextArea/index.jsx [hunk 19]
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable react/jsx-props-no-spreading */
```
Comment (Mahmoudzaid2994) [inline src/components/shared/FormTextArea/index.jsx:1]:
```suggestion
```

File: src/components/shared/FilesList/index.jsx (hunk 11 above)
Comment (hosamAjlouni-op) [reply src/components/shared/FilesList/index.jsx:16]: already discussed this with @AhmadAlShwaiki and sattled on this, 
another way is to generate a unique id for each selected file (image), insights?

**Review by Mahmoudzaid2994 — APPROVED** [review]

**Pull request merged**